File: android/src/main/kotlin/com/qusadprod/sing_box/BypassManager.kt
package com.qusadprod.sing_box

import android.content.Context
import android.content.SharedPreferences
import com.google.gson.Gson
import com.google.gson.reflect.TypeToken
import java.util.concurrent.CopyOnWriteArraySet

/**
 * Manager for managing bypass lists
 * Stores lists of applications, domains, and subnets
 */
class BypassManager(private val context: Context) {
    private val prefs: SharedPreferences = context.getSharedPreferences("sing_box_bypass", Context.MODE_PRIVATE)
    private val gson = Gson()
    
    // Bypass lists in memory for quick access
    private val bypassApps = CopyOnWriteArraySet<String>()
    private val bypassDomains = CopyOnWriteArraySet<String>()
    private val bypassSubnets = CopyOnWriteArraySet<String>()
    
    init {
        loadFromPreferences()
    }
    
    // ========== Bypass Apps ==========
    
    fun addAppToBypass(packageName: String): Boolean {
        if (bypassApps.add(packageName)) {
            saveAppsToPreferences()
            return true
        }
        return false
    }
    
    fun removeAppFromBypass(packageName: String): Boolean {
        if (bypassApps.remove(packageName)) {
            saveAppsToPreferences()
            return true
        }
        return false
    }
    
    fun getBypassApps(): List<String> {
        return bypassApps.toList()
    }
    
    // ========== Bypass Domains ==========
    
    fun addDomainToBypass(domain: String): Boolean {
        if (bypassDomains.add(domain)) {
            saveDomainsToPreferences()
            return true
        }
        return false
    }
    
    fun removeDomainFromBypass(domain: String): Boolean {
        if (bypassDomains.remove(domain)) {
            saveDomainsToPreferences()
            return true
        }
        return false
    }
    
    fun getBypassDomains(): List<String> {
        return bypassDomains.toList()
    }
    
    // ========== Bypass Subnets ==========
    
    fun addSubnetToBypass(subnet: String): Boolean {
        if (isValidSubnet(subnet) && bypassSubnets.add(subnet)) {
            saveSubnetsToPreferences()
            return true
        }
        return false
    }
    
    fun removeSubnetFromBypass(subnet: String): Boolean {
        if (bypassSubnets.remove(subnet)) {
            saveSubnetsToPreferences()
            return true
        }
        return false
    }
    
    fun getBypassSubnets(): List<String> {
        return bypassSubnets.toList()
    }
    
    // ========== Helper Methods ==========
    
    private fun isValidSubnet(subnet: String): Boolean {
        // Simple CIDR format check (e.g., "192.168.1.0/24")
        return subnet.matches(Regex("^\\d{1,3}\\.\\d{1,3}\\.\\d{1,3}\\.\\d{1,3}/\\d{1,2}$"))
    }
    
    private fun loadFromPreferences() {
        // Load apps
        val appsJson = prefs.getString("bypass_apps", "[]")
        val appsType = object : TypeToken<List<String>>() {}.type
        val appsList: List<String> = gson.fromJson(appsJson, appsType) ?: emptyList()
        bypassApps.addAll(appsList)
        
        // Load domains
        val domainsJson = prefs.getString("bypass_domains", "[]")
        val domainsType = object : TypeToken<List<String>>() {}.type
        val domainsList: List<String> = gson.fromJson(domainsJson, domainsType) ?: emptyList()
        bypassDomains.addAll(domainsList)
        
        // Load subnets
        val subnetsJson = prefs.getString("bypass_subnets", "[]")
        val subnetsType = object : TypeToken<List<String>>() {}.type
        val subnetsList: List<String> = gson.fromJson(subnetsJson, subnetsType) ?: emptyList()
        bypassSubnets.addAll(subnetsList)
    }
    
    private fun saveAppsToPreferences() {
        val appsJson = gson.toJson(bypassApps.toList())
        prefs.edit().putString("bypass_apps", appsJson).apply()
    }
    
    private fun saveDomainsToPreferences() {
        val domainsJson = gson.toJson(bypassDomains.toList())
        prefs.edit().putString("bypass_domains", domainsJson).apply()
    }
    
    private fun saveSubnetsToPreferences() {
        val subnetsJson = gson.toJson(bypassSubnets.toList())
        prefs.edit().putString("bypass_subnets", subnetsJson).apply()
    }
}

File: ios/Classes/Managers/BypassManager.swift
import Foundation

// Manager for managing bypass lists (apps, domains, subnets)
// Analog of Android BypassManager
public class BypassManager {
    private let userDefaults: UserDefaults
    private let bypassAppsKey = "sing_box_bypass_apps"
    private let bypassDomainsKey = "sing_box_bypass_domains"
    private let bypassSubnetsKey = "sing_box_bypass_subnets"
    
    // Bypass lists in memory for quick access
    private var bypassApps: Set<String> = []
    private var bypassDomains: Set<String> = []
    private var bypassSubnets: Set<String> = []
    
    public init() {
        // Use App Group for access from Extension
        if let groupDefaults = UserDefaults(suiteName: FilePath.groupName) {
            self.userDefaults = groupDefaults
        } else {
            self.userDefaults = UserDefaults.standard
        }
        loadFromUserDefaults()
    }
    
    // ========== Bypass Apps ==========
    
    public func addAppToBypass(_ bundleId: String) -> Bool {
        if bypassApps.insert(bundleId).inserted {
            saveAppsToUserDefaults()
            return true
        }
        return false
    }
    
    public func removeAppFromBypass(_ bundleId: String) -> Bool {
        if bypassApps.remove(bundleId) != nil {
            saveAppsToUserDefaults()
            return true
        }
        return false
    }
    
    public func getBypassApps() -> [String] {
        return Array(bypassApps)
    }
    
    // ========== Bypass Domains ==========
    
    public func addDomainToBypass(_ domain: String) -> Bool {
        let normalizedDomain = normalizeDomain(domain)
        if normalizedDomain.isEmpty {
            return false
        }
        if bypassDomains.insert(normalizedDomain).inserted {
            saveDomainsToUserDefaults()
            return true
        }
        return false
    }
    
    public func removeDomainFromBypass(_ domain: String) -> Bool {
        let normalizedDomain = normalizeDomain(domain)
        if bypassDomains.remove(normalizedDomain) != nil {
            saveDomainsToUserDefaults()
            return true
        }
        return false
    }
    
    public func getBypassDomains() -> [String] {
        return Array(bypassDomains)
    }
    
    // ========== Bypass Subnets ==========
    
    public func addSubnetToBypass(_ subnet: String) -> Bool {
        if isValidSubnet(subnet) && bypassSubnets.insert(subnet).inserted {
            saveSubnetsToUserDefaults()
            return true
        }
        return false
    }
    
    public func removeSubnetFromBypass(_ subnet: String) -> Bool {
        if bypassSubnets.remove(subnet) != nil {
            saveSubnetsToUserDefaults()
            return true
        }
        return false
    }
    
    public func getBypassSubnets() -> [String] {
        return Array(bypassSubnets)
    }
    
    // ========== Helper Methods ==========
    
    private func isValidSubnet(_ subnet: String) -> Bool {
        // Simple CIDR format check (e.g., "192.168.1.0/24")
        let pattern = "^\\d{1,3}\\.\\d{1,3}\\.\\d{1,3}\\.\\d{1,3}/\\d{1,2}$"
        return subnet.range(of: pattern, options: .regularExpression) != nil
    }
    
    private func normalizeDomain(_ domain: String) -> String {
        var result = domain.lowercased()
        
        // Remove protocol
        result = result.replacingOccurrences(of: "^https?://", with: "", options: .regularExpression)
        
        // Remove "www."
        result = result.replacingOccurrences(of: "^www\\.", with: "", options: .regularExpression)
        
        // Remove everything after first slash
        if let slashIndex = result.firstIndex(of: "/") {
            result = String(result[..<slashIndex])
        }
        
        return result.trimmingCharacters(in: .whitespaces)
    }
    
    private func loadFromUserDefaults() {
        // Load apps
        if let appsData = userDefaults.data(forKey: bypassAppsKey),
           let apps = try? JSONDecoder().decode([String].self, from: appsData) {
            bypassApps = Set(apps)
        }
        
        // Load domains
        if let domainsData = userDefaults.data(forKey: bypassDomainsKey),
           let domains = try? JSONDecoder().decode([String].self, from: domainsData) {
            bypassDomains = Set(domains)
        }
        
        // Load subnets
        if let subnetsData = userDefaults.data(forKey: bypassSubnetsKey),
           let subnets = try? JSONDecoder().decode([String].self, from: subnetsData) {
            bypassSubnets = Set(subnets)
        }
    }
    
    private func saveAppsToUserDefaults() {
        if let data = try? JSONEncoder().encode(Array(bypassApps)) {
            userDefaults.set(data, forKey: bypassAppsKey)
        }
    }
    
    private func saveDomainsToUserDefaults() {
        if let data = try? JSONEncoder().encode(Array(bypassDomains)) {
            userDefaults.set(data, forKey: bypassDomainsKey)
        }
    }
    
    private func saveSubnetsToUserDefaults() {
        if let data = try? JSONEncoder().encode(Array(bypassSubnets)) {
            userDefaults.set(data, forKey: bypassSubnetsKey)
        }
    }
}

